src/executor/cursor.rs

```diff
@@ -0,0 +1,264 @@
+use crate::{
+    ColumnTrait, ConnectionTrait, DbErr, EntityTrait, FromQueryResult, Iterable,
+    PrimaryKeyToColumn, PrimaryKeyTrait, Select, SelectModel, SelectorTrait,
+};
+use sea_query::{Condition, IntoValueTuple, Order, SelectStatement, SimpleExpr, Value};
+use std::marker::PhantomData;
+
+/// Limit number of rows to be fetched from the cursor
+#[derive(Debug)]
+pub enum CursorLimit {
+    /// Fetch first N rows in ascending order of primary key
+    First(u32),
+    /// Fetch last N rows in descending order of primary key
+    Last(u32),
+}
+
+/// Cursor pagination
+#[derive(Debug)]
+pub struct Cursor<E, S>
+where
+    E: EntityTrait,
+    S: SelectorTrait,
+{
+    pub(crate) query: SelectStatement,
+    pub(crate) condition: Option<Condition>,
+    pub(crate) limit: Option<CursorLimit>,
+    pub(crate) phantom: PhantomData<(E, S)>,
+}
+
+impl<E, S> Cursor<E, S>
+where
+    E: EntityTrait,
+    S: SelectorTrait,
+{
+    /// Initialize a cursor
+    pub fn new(query: SelectStatement) -> Self {
+        Self {
+            query,
+            condition: None,
+            limit: None,
+            phantom: PhantomData,
+        }
+    }
+
+    /// Filter rows with primary key value less than the input value
+    pub fn before(&mut self, values: <E::PrimaryKey as PrimaryKeyTrait>::ValueType) -> &mut Self {
+        self.condition = Some(get_condition::<E, _>(values, |c: E::Column, v: Value| {
+            c.lt(v)
+        }));
+        self
+    }
+
+    /// Filter rows with primary key value greater than the input value
+    pub fn after(&mut self, values: <E::PrimaryKey as PrimaryKeyTrait>::ValueType) -> &mut Self {
+        self.condition = Some(get_condition::<E, _>(values, |c: E::Column, v: Value| {
+            c.gt(v)
+        }));
+        self
+    }
+
+    /// Limit result set to only first N rows in ascending order of the primary key
+    pub fn first(&mut self, num_rows: u32) -> &mut Self {
+        self.limit = Some(CursorLimit::First(num_rows));
+        self
+    }
+
+    /// Limit result set to only last N rows in ascending order of the primary key
+    pub fn last(&mut self, num_rows: u32) -> &mut Self {
+        self.limit = Some(CursorLimit::Last(num_rows));
+        self
+    }
+
+    /// Fetch the rows
+    pub async fn all<C>(&self, db: &C) -> Result<Vec<S::Item>, DbErr>
+    where
+        C: ConnectionTrait,
+    {
+        let mut query = self.query.clone();
```

Comment:
That clone is somewhat annoying since most of the time self is not reused.

Reply:
Managed to avoid the clone :)